currency/rate_engines.go

```diff
@@ -0,0 +1,42 @@
+package currency
+
+import "strings"
+
+const (
+	CustomRates = iota
+	PBSRates
+)
+
+type RateEngines struct {
+	conversions []Conversions
+}
+
+// NewRates creates a new Rates object holding currencies rates
+func NewRateEngines(customRates, pbsRates Conversions) *RateEngines {
+	return &RateEngines{
+		conversions: []Conversions{customRates, pbsRates},
+	}
+}
+
+// GetRate returns the conversion rate between two currencies prioritizing
+// the customRates currency rate over that of the PBS currency rate service
+// returns an error if both Conversions objects return error.
+func (re *RateEngines) GetRate(from string, to string) (float64, error) {
+
+	rate, err := re.conversions[CustomRates].GetRate(from, to)
+	if err == nil || !strings.HasPrefix(err.Error(), `Currency conversion rate not found`) {
+		// valid custom conversion rate was found, return this
+		// value because custom rates take priority over PBS rates
+		return rate, err
+	}
+
+	// because the custom rates' GetRate() call returned an error other than "conversion
+	// rate not found", there's nothing wrong with the 3 letter currency code so let's
+	// try the PBS rates instead
+	return re.conversions[PBSRates].GetRate(from, to)
+}
+
+// No need to call GetRates on RateEngines
```

Comment:
Why?

Reply:
This function is only ever used (outside of test cases) to put up the `ConverterInfo` the `/currency/rates` endpoint returns. At the `exchange` level, to return a map with all of the currencies does not seem useful. A way to implement it, would be to copy the PBS rates to a new map and overwrite the custom rates on top (something similar to what a previous version of the `getAuctionCurrencyRates(requestRates *openrtb_ext.ExtRequestCurrency)` did). Under the new `RatesEngines` approach I see no need. I don't think it would ever be used. Thoughts?

I can still implement it if you think we need to:
```
// Implementing GetRates on RateEngines
func (re *RateEngines) GetRates() *map[string]map[string]float64 {
	serverRatesMap := re.conversions[PBSRates].GetRates()
	customRatesMap := re.conversions[CustomRates].GetRates()

	returnRatesMap := make(map[string]map[string]float64, len(*serverRatesMap))

	//Copy server rates into returnRatesMap
	for fromCurrency, rates := range (*serverRatesMap) {
		returnRatesMap[fromCurrency] = make(map[string]float64, len(rates))
		for toCurrency, rate := range rates {
			returnRatesMap[fromCurrency][toCurrency] = rate
		}
	}

	//Overwrite with customRates
	for fromCurrency, rates := range (*customRatesMap) {
		if _, defined := returnRatesMap[fromCurrency]; !defined {
			returnRatesMap[fromCurrency] = make(map[string]float64, len(rates))
		}

		for toCurrency, rate := range rates {
			returnRatesMap[fromCurrency][toCurrency] = rate

			// Write entries to inverse of conversion rates if needed
			if _, inverseIncluded := customRates.ConversionRates[toCurrency][fromCurrency]; !inverseIncluded {
				if _, defined := returnRatesMap[toCurrency]; !defined {
					returnRatesMap[toCurrency] = map[string]float64{fromCurrency: 1.00 / rate}
				} else {
					returnRatesMap[toCurrency][fromCurrency] = 1.00 / rate
				}
			}
		}
	}

	//Return pointer
	return *returnRatesMap
}
```

Reply:
> This function is only ever used (outside of test cases) to put up the `ConverterInfo` the `/currency/rates` endpoint returns.

Ah, ok. I understand. Seems fine for now. I think we should revisit this design in the future such that the `GetRate` for the bidder loop doesn't have a dependency on the `GetRates` for the currency conversion endpoint.